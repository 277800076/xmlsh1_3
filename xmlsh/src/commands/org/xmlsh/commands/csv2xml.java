/**
 * $Id: xpwd.java 21 2008-07-04 08:33:47Z daldei $
 * $Date: 2008-07-04 04:33:47 -0400 (Fri, 04 Jul 2008) $
 *
 */

package org.xmlsh.commands;

import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.Reader;
import java.util.List;

import javax.xml.transform.sax.TransformerHandler;

import org.xml.sax.Attributes;
import org.xml.sax.SAXException;
import org.xml.sax.helpers.AttributesImpl;
import org.xmlsh.commands.util.CSVParser;
import org.xmlsh.commands.util.CSVRecord;
import org.xmlsh.core.Options;
import org.xmlsh.core.XCommand;
import org.xmlsh.core.XEnvironment;
import org.xmlsh.core.XValue;
import org.xmlsh.util.Util;

/*
 * 
 * Convert CSV files to an XML file
 * Arguments
 * 	-root		root element (default "root")
 *  -row		row	 element (default "row")
 *  -col		col	 element (defauilt "col")
 *  -header		read first row for header names
 *  -attr		write in attribute normal format
 */

public class csv2xml extends XCommand
{

	public static void main( String args[] ) throws Exception
	{
		csv2xml cmd = new csv2xml();

		cmd.run( args );
		
		
		
	}

	
	
	public int run(  List<XValue> args , XEnvironment env )	throws Exception
	{

		

		Options opts = new Options( "root:,row:,col:,header,attr,encoding:,delim:" , args );
		opts.parse();
		
		// root node
		String root = opts.getOptString("root", "root");
		String row = opts.getOptString("row", "row");
		String col = opts.getOptString("col", "col");
		String delim = opts.getOptString("delim", ",");
		String encoding = opts.getOptString("encoding", "Cp1252");
		boolean bHeader = opts.hasOpt("header");
		boolean bAttr = opts.hasOpt("attr");
		
		List<XValue> xvargs = opts.getRemainingArgs();
		
// Output XML
		OutputStream stdout = env.getStdout();
		TransformerHandler hd = Util.getTransformerHander(stdout);
		hd.startDocument();
		Attributes attrs = new AttributesImpl();
		hd.startElement("", root,root,attrs);
		
// Input is stdin and/or list of commands
		
		InputStream in = null;
		if( xvargs.size() == 0 || xvargs.get(0).toString().equals("-"))
			in = env.getStdin();
		else
			in = new FileInputStream(  env.getShell().getFile(xvargs.get(0)) );
		
		
		Reader ir = new InputStreamReader( in , encoding );
		CSVParser parser = new CSVParser( delim.charAt(0) );
		
		CSVRecord header = null ;
		if( bHeader ){
			String line = readLine(ir);
			if( line != null )
				header = parser.parseLine(line);
		}
		
		String line;
		while( (line = readLine(ir)) != null ){
			CSVRecord csv = parser.parseLine(line);
			addElement( hd , csv , row , col , bAttr , header );
		}
		hd.endElement("", root,root);
		hd.endDocument();
		
		ir.close();
		
		
		
		return 0;
		
	}



	private void addElement(
		TransformerHandler hd, 
		CSVRecord csv ,
		String row, 
		String col, 
		boolean battr,
		CSVRecord header) throws SAXException 
	{
		
		
		// Attribute normal format
		if( battr ){
			AttributesImpl attrs = new AttributesImpl();
			for( int i = 0 ; i < csv.getNumFields() ; i++ ){
				String name = getAttrName( i , col , header );
				attrs.addAttribute("", name , name, "CDATA", csv.getField(i));
			}
			hd.startElement("", row,row,attrs);
			hd.endElement("", row, row);
		} else {
			AttributesImpl attrs = new AttributesImpl();
			hd.startElement("", row,row,attrs);

			for( int i = 0 ; i < csv.getNumFields() ; i++ ){
				String name = getColName( i , col , header );
				hd.startElement("", name, name, attrs);
				char[] chars = csv.getField(i).toCharArray();
				hd.characters(chars,0,chars.length);
				hd.endElement("", name, name);
			}
			hd.endElement("", row,row);
		}
	}


	private String getColName(int i, String col, CSVRecord header) {
		if( header != null && header.getNumFields() > i )
			return toXmlName( header.getField(i));
		else
			return col ;
	}



	// Get an attribute name 
	private String getAttrName(int i, String col, CSVRecord header) {
		if( header != null && header.getNumFields() > i )
			return toXmlName( header.getField(i));
		else
			return col + (i + 1)  ;
		
	}



	private String toXmlName(String field) {
		return field.replaceAll("[^a-zA-Z0-9_]","-");
	}



	private String readLine(Reader ir) throws IOException {
		return Util.readLine(ir);
	}
	
}

//
//
//Copyright (C) 2008, David A. Lee.
//
//The contents of this file are subject to the "Simplified BSD License" (the "License");
//you may not use this file except in compliance with the License. You may obtain a copy of the
//License at http://www.opensource.org/licenses/bsd-license.php 
//
//Software distributed under the License is distributed on an "AS IS" basis,
//WITHOUT WARRANTY OF ANY KIND, either express or implied.
//See the License for the specific language governing rights and limitations under the License.
//
//The Original Code is: all this file.
//
//The Initial Developer of the Original Code is David A. Lee
//
//Portions created by (your name) are Copyright (C) (your legal entity). All Rights Reserved.
//
//Contributor(s): none.
//
