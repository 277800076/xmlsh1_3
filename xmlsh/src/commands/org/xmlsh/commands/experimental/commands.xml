<?xml version="1.0"?>
<commands>

<command name="xml2json" url="http://www.xmlsh.org/CommandXmlsh2json"/>
<command name="xobsufcate" url="http://www.xmlsh.org/CommandXobsufcate"/>
<command name="xsysinfo" url="http://www.xmlsh.org/CommandXsysinfo"/>
</commands>