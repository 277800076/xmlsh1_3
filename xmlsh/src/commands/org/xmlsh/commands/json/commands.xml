<commands>
 
   <function name="boolean" url="http://www.xmlsh.org/FunctionJsonBoolean">
		    <synopsis>construct a JSON boolean</synopsis>
		    </usage>
	  </function>
	  
	     <function name="null" url="http://www.xmlsh.org/FunctionJsonNull">
		    <synopsis>construct a JSON null</synopsis>
		    </usage>
	  </function>
	     <function name="array" url="http://www.xmlsh.org/FunctionJsonarray">
		    <synopsis>construct a JSON array</synopsis>
		    </usage>
	  </function>
	     <function name="json" url="http://www.xmlsh.org/FunctionJsonJson">
		    <synopsis>construct a JSON object from a string</synopsis>
		    </usage>
	  </function>
	     <function name="object" url="http://www.xmlsh.org/FunctionJsonObject">
		    <synopsis>construct a JSON object</synopsis>
		    </usage>
	  </function>
	     <function name="boolean" url="http://www.xmlsh.org/FunctionJsonPath">
		    <synopsis>Evaluate a JsonPath on a JSON object</synopsis>
		    </usage>
	  </function>
	  	     <function name="string" url="http://www.xmlsh.org/FunctionJsonString">
		    <synopsis>construct a JSON string</synopsis>
		    </usage>
	  </function>
	  	     <function name="value" url="http://www.xmlsh.org/FunctionJsonValue">
		    <synopsis>construct a JSON value</synopsis>
		    </usage>
	  </function>
</commands>
