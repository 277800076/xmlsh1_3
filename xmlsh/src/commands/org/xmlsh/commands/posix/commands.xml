<?xml version="1.0"?>
<commands>

<command name="cat" url="http://www.xmlsh.org/CommandCat">
	<snopsis>copy or concatenate files</snopsis>
</command>
<command name="chmod" url="http://www.xmlsh.org/CommandChmod">
	<synopsis>Change file permissions</synopsis>
</command>
<command name="cp" url="http://www.xmlsh.org/CommandCp">
	<synopsis>copy one or more files</synopsis>
</command>
<command name="ls" url="http://www.xmlsh.org/CommandLs">
	<synopsis>list files in a directory</synopsis>
</command>
<command name="mkdir" url="http://www.xmlsh.org/CommandMkdir">
	<synopsis>Create directories</synopsis>
</command>
<command name="more" url="http://www.xmlsh.org/CommandMore">
	<synopsis>Simple pagenator</synopsis>
</command>
<command name="mv" url="http://www.xmlsh.org/CommandMv">
	<synopsis>Move/Rename files</synopsis>
</command>
<command name="rm" url="http://www.xmlsh.org/CommandRm">
	<synopsis>Remove/Delete files</synopsis>
</command>
<command name="rmdir" url="http://www.xmlsh.org/CommandRmdir">
	<synopsis>Remove/Delete directories</synopsis>
</command>
<command name="sleep" url="http://www.xmlsh.org/CommandSleep">
	<synopsis>Delay execution</synopsis>
</command>
<command name="touch" url="http://www.xmlsh.org/CommandTouch" >
	<synopsis>create and update timestamp of a file</synopsis>
</command>

</commands>
	