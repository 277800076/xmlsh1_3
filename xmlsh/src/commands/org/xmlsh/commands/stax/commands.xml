
<commands>

	<command name="closeReader" url="http://www.xmlsh.org/CommandStAXcloseReader">
		<synopsis>close a StAX reader</synopsis>
		<usage>
			<para>closeReader $reader</para>
		</usage>
	</command>

	<function name="getAttribute" url="http://www.xmlsh.org/FuncitonStAXgetAttribute">
		<synopsis>Gets an attribute value from a StAX START_ELEMENT Event</synopsis>
		<usage>
			<para>getAttribute( $event qname  )</para>
		</usage>
	</function>

	<function name="getData" url="http://www.xmlsh.org/FunctionStAXgetData">
		<synopsis>Gets the Data of a StAX Event</synopsis>
		<usage>
			<para>getData($event)</para>
		</usage>
	</function>

	<function name="getEventType" url="http://www.xmlsh.org/FunctionStAXgetEventType">
		<synopsis>Gets event type of a StAX Event</synopsis>
		<usage>
			<para>getEventType($event)</para>
		</usage>
	</function>
	<function name="getName" url="http://www.xmlsh.org/FunctionStAXgetName">
		<synopsis>Gets name of a StAX Event as a QName</synopsis>
		<usage>
			<para>getName($event)</para>
		</usage>
	</function>
	<function name="getNamespace" url="http://www.xmlsh.org/FunctionStAXgetNamespace">
		<synopsis>Gets namespace URI of a StAX Event as a String</synopsis>
		<usage>
			<para>getNamespace($event)</para>
		</usage>
	</function>

	<function name="hasNext" url="http://www.xmlsh.org/FunctionStAXhasNext">
		<synopsis>Advances the StAX Stream to the next matching event</synopsis>
		<usage>
			<para>hasNext($reader)</para>
			<para>hasNext($reader event_type)</para>
			<para>hasNext($reader event_type qname)</para>
		</usage>
	</function>

	<function name="nextEvent" url="http://www.xmlsh.org/FunctionStAXnextEvent">
		<synopsis>Advances the StAX Stream to the next matching event</synopsis>
		<usage>
			<para>nextEvent($reader)</para>
			<para>nextEvent($reader event_type)</para>
			<para>nextEvent($reader event_type qname)</para>
		</usage>
	</function>
	<function name="newEventReader" url="http://www.xmlsh.org/FunctionStAXnewEventReader">
		<synopsis>Creates a new XMLEventReader</synopsis>
		<usage>
			<para>newEventReader()</para>
			<para>newEventReader(uri)</para>
		</usage>
	</function>
</commands>