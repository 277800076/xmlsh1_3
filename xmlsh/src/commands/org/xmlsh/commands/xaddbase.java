/**
 * $Id: $
 * $Date: $
 *
 */

package org.xmlsh.commands;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.List;

import javax.xml.transform.Source;

import net.sf.saxon.AugmentedSource;
import net.sf.saxon.event.Builder;
import net.sf.saxon.om.Axis;
import net.sf.saxon.om.AxisIterator;
import net.sf.saxon.om.DocumentInfo;
import net.sf.saxon.om.Item;
import net.sf.saxon.om.MutableNodeInfo;
import net.sf.saxon.om.NamePool;
import net.sf.saxon.om.NodeInfo;
import net.sf.saxon.om.StandardNames;
import net.sf.saxon.s9api.DocumentBuilder;
import net.sf.saxon.s9api.Processor;
import net.sf.saxon.s9api.SaxonApiException;
import net.sf.saxon.s9api.Serializer;
import net.sf.saxon.s9api.XdmItem;
import net.sf.saxon.s9api.XdmNode;
import net.sf.saxon.s9api.XdmNodeKind;
import net.sf.saxon.tree.DocumentImpl;
import net.sf.saxon.type.Type;
import org.xmlsh.core.Options;
import org.xmlsh.core.OutputPort;
import org.xmlsh.core.XCommand;
import org.xmlsh.core.XValue;
import org.xmlsh.core.Options.OptionValue;
import org.xmlsh.sh.shell.Shell;
import org.xmlsh.util.Util;


public class xaddbase extends XCommand {

	private DocumentBuilder mBuilder;
	private Processor mProcessor;

	private void setupBuilders()
	{
/*		
		mProcessor = new Processor(false);
		mProcessor.setConfigurationProperty(FeatureKeys.TREE_MODEL, net.sf.saxon.event.Builder.LINKED_TREE);
*/
		mProcessor = Shell.getProcessor();
		

		mBuilder = mProcessor.newDocumentBuilder();
		
	}
	
	
	
	@Override
	public int run( List<XValue> args )
	throws Exception 
	{
		boolean		opt_all		= false ;
		boolean		opt_relative = false ;
		
		
		Options opts = new Options( "i:,a=all,r=relative" , args );
		opts.parse();
		
		setupBuilders();

		XdmNode	context = null;

		
		OptionValue ov = opts.getOpt("i");

			
		// If -i argument is an XML expression take the first node as the context
		if( ov != null  && ov.getValue().isXExpr() ){
			XdmItem item = ov.getValue().asXdmValue().itemAt(0);
			if( item instanceof XdmNode )
			//   context = (XdmNode) item ; // builder.build(((XdmNode)item).asSource());
			 // context = (XdmNode) ov.getValue().toXdmValue();
			context = importNode( (XdmNode)item);

		}
		if( context == null )
		{

			if( ov != null && ! ov.getValue().toString().equals("-"))
				context = build( getSource(ov.getValue()));
			else {
				context = build(getStdin().asSource());
			}	
		}
		
		opt_all 	= opts.hasOpt("all");
		opt_relative = opts.hasOpt("relative");
		add_base( context , opt_all , opt_relative );
			
		


		OutputPort stdout = getStdout();
		Util.writeXdmValue( context , stdout.asDestination() );
		stdout.writeSequenceTerminator();
		
		return 0;

	}
	
	


	private void add_child_base(NodeInfo parent, boolean opt_all, boolean opt_relative, String parentBaseURI ) throws URISyntaxException 
	{
		
		
		AxisIterator iter = parent.iterateAxis( Axis.CHILD );

		
		Item item;
		while( (item = iter.next()) != null ){

			if( item instanceof MutableNodeInfo  ){
				MutableNodeInfo node = (MutableNodeInfo) item;
				if( node.getNodeKind() == Type.ELEMENT ){
					String baseURI = node.getBaseURI();
					if( opt_all ||  ! Util.isEqual( baseURI , parentBaseURI ) )
						addAttribute( node , "xml" , "http://www.w3.org/XML/1998/namespace", "base" , resolve(parentBaseURI , baseURI, opt_relative) );
					else
						removeAttribute( node , "xml" , "http://www.w3.org/XML/1998/namespace", "base");
					
					add_child_base( node , opt_all , opt_relative, baseURI );
					
				}
			}
			
			
			
		}
		
	}

	
	
	
	


	private String resolve(String parentBaseURI, String baseURI, boolean opt_relative) throws URISyntaxException {
		if(! opt_relative )
			return baseURI ;
		
		URI u = new URI( parentBaseURI );
		// Remove final path component from parent 
		String path = u.getPath();
		int slash = path.lastIndexOf('/');
		if( slash >= 0 )
			path = path.substring(0,slash);
		URI parent_URI = 
			new URI(u.getScheme(),
			        u.getUserInfo(), u.getHost(), u.getPort(),
			        path, 
			        u.getQuery(),
			        u.getFragment());

		
		URI base_URI = new URI( baseURI );
		
		URI relative = parent_URI.relativize(base_URI);
		return relative.toString();
	}



	private void add_base(XdmNode xroot, boolean opt_all, boolean opt_relative ) throws URISyntaxException 
	{
	
		
		NodeInfo root = null;
		if( xroot.getNodeKind() == XdmNodeKind.DOCUMENT )
			root = ((DocumentImpl) xroot.getUnderlyingNode()).getDocumentElement();
		else
			root = xroot.getUnderlyingNode();
		
		
		MutableNodeInfo node = (MutableNodeInfo) root;
		if( node.getNodeKind() == Type.ELEMENT ){
			String baseURI = node.getBaseURI();
			addAttribute( node , "xml" , "http://www.w3.org/XML/1998/namespace", "base" , baseURI );
			add_child_base( node , opt_all , opt_relative, baseURI );
			
			
		}

		
	}


	private void addAttribute(MutableNodeInfo node, String prefix, String uri , String local , String value ) {

		NamePool pool = node.getNamePool();
		int nameCode  = pool.allocate(prefix , uri , local  );
		node.putAttribute(nameCode,  StandardNames.XS_UNTYPED_ATOMIC, value , 0);

	}


	private void removeAttribute(MutableNodeInfo node, String prefix, String uri, String local) {
		NamePool pool = node.getNamePool();
		int nameCode  = pool.allocate(prefix , uri , local  );
		node.removeAttribute(nameCode);
		
	}




	/*
	 * Import the node using the builder into this object model
	 */
	private XdmNode importNode( XdmNode node ) throws SaxonApiException
	{
		Source src = node.asSource();
		return build(src);
	}

	

	
	/*
	 * Creates/Builds a Tree (LINKED_TREE) type node from any source
	 */
	
	private XdmNode build( Source src ) throws SaxonApiException
	{
		// @TODO: To get over a bug in Saxon's build() have to use the root element
		// instead of a document node to force building of a linked tree model
		// Otherwise the source is just returned unchnaged
		
		if( src instanceof DocumentInfo  )
			src = (NodeInfo)(((DocumentInfo)src).iterateAxis(net.sf.saxon.om.Axis.CHILD).next());

		
		AugmentedSource asrc = AugmentedSource.makeAugmentedSource(src); 
		asrc.setTreeModel(Builder.LINKED_TREE); 
		return mBuilder.build(asrc);
		
	}

	

}

//
//
//Copyright (C) 2008, David A. Lee.
//
//The contents of this file are subject to the "Simplified BSD License" (the "License");
//you may not use this file except in compliance with the License. You may obtain a copy of the
//License at http://www.opensource.org/licenses/bsd-license.php 
//
//Software distributed under the License is distributed on an "AS IS" basis,
//WITHOUT WARRANTY OF ANY KIND, either express or implied.
//See the License for the specific language governing rights and limitations under the License.
//
//The Original Code is: all this file.
//
//The Initial Developer of the Original Code is David A. Lee
//
//Portions created by (your name) are Copyright (C) (your legal entity). All Rights Reserved.
//
//Contributor(s): none.
//
