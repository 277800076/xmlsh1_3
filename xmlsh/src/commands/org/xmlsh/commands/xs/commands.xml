<commands>
 
   <function name="string" url="http://www.xmlsh.org/FunctionXSString">
		    <synopsis>Convert to string</synopsis>
		    <usage>
			      <para>string( args )</para>
		    </usage>
	  </function>
   
   <function name="integer" url="http://www.xmlsh.org/FunctionXSInteger">
		    <synopsis>Convert to integer</synopsis>
		    <usage>
			      <para>integer( arg )</para>
		    </usage>
	  </function>
	  
	  
	  
</commands>
