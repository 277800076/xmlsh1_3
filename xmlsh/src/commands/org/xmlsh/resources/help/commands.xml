<commands>
   <command name=":" url="http://www.xmlsh.org/CommandColon">
		    <synopsis>the null command</synopsis>
		    <usage>
			      <para>: ignored arguments</para>
		    </usage>
	  </command>
   <command name="[" url="http://www.xmlsh.org/CommandTest">
   	<synopsis>Evaluates expressions</synopsis>
	</command>

   <command name="break" url="http://www.xmlsh.org/CommandBreak">
   <synopsis>Exits one level of a while, for or until</synopsis>
   </command>
   <command name="cd" url="http://www.xmlsh.org/CommandCd">
   <synopsis>change or print current directory</synopsis>
   </command>

   <command name="continue" url="http://www.xmlsh.org/CommandContinue">
	   <synopsis>Continues the iteration of one level of a loop</synopsis>
   </command>


   <command name="csv2xml" url="http://www.xmlsh.org/CommandCsv2xml">
	   <synopsis>convert csv file to xml</synopsis>
   </command>


   <command name="declare" url="http://www.xmlsh.org/CommandDeclare">
		    <synopsis>declare Declares or lists namespaces</synopsis>
		    <usage>
			      <para>declare</para>
			      <para>declare namespace [prefix=[uri]]</para>
		    </usage>
	  </command>
   <command name="echo" url="http://www.xmlsh.org/CommandEcho">
		    <synopsis>prints arguments as text</synopsis>
	  </command>
   <command name="eval" url="http://www.xmlsh.org/CommandEval">
		    <synopsis>Evaluate a shell command</synopsis>
	  </command>
   <command name="exit" url="http://www.xmlsh.org/CommandExit">
		    <synopsis>exit the shell, script, or function</synopsis>
	  </command>



   <command name="false" url="http://www.xmlsh.org/CommandFalse">
		    <synopsis>returns false</synopsis>
	  </command>
   <command name="fixed2xml" url="http://www.xmlsh.org/CommandFixed2xml">
    <synopsis>convert fixed field files to xml</synopsis>

   </command>
   <command name="help" url="http://www.xmlsh.org/CommandHelp">
		    <synopsis>help for xlmsh commands</synopsis>
		    <usage>
			      <para>help [-n] commands</para>
			      <para>help [-n] command</para>
		    </usage>
		    <options>
			      <option>
				        <arg>-n</arg>
				        <para>Do not launch a browser for details</para>
			      </option>
			      <option>
				        <arg>-xml</arg>
				        <para>Outputs the xml format</para>
			      </option>
		    </options>
	  </command>
	 <command name="http" url="http://www.xmlsh.org/CommandHttp">
	 	<synopsis>Http client command</synopsis>
	 </command>
	<command name="httpserver" url="http://www.xmlsh.org/CommandHttpserver">
		<synopsis>Http server</synopsis>
	</command>
   <command name="import" url="http://www.xmlsh.org/CommandImport">
   
       <synopsis>Import modules, packages and commands</synopsis>

   </command>
   <command name="jcall" url="http://www.xmlsh.org/CommandJcall">
        <synopsis>call java from within the same JVM</synopsis>   
   </command>
   <command name="jobs" url="http://www.xmlsh.org/CommandJobs">
         <synopsis>prints jobs</synopsis>   
   </command>
   <command name="log" url="http://www.xmlsh.org/CommandLog">
		<synopsis>Writes to log4j log</synopsis>   
   </command>
   <command name="read" url="http://www.xmlsh.org/CommandRead">
   		<synopsis>reads text into a variable</synopsis>
   </command>
   <command name="require" url="http://www.xmlsh.org/CommandRequire">
   		<synopsis>Checks the current shell for a specified minimum version</synopsis>
   </command>
   <command name="rngvalidate" url="http://www.xmlsh.org/CommandRngvalidate">
   		<synopsis>Validates with an RNG schema</synopsis>
   </command>
   <command name="set" url="http://www.xmlsh.org/CommandSet">
   		<synopsis>set positional parameters and view variables</synopsis>
   </command>
   <command name="shift" url="http://www.xmlsh.org/CommandShift">
   		<synopsis>shift positional parameters</synopsis>
   </command>
   <command name="source" url="http://www.xmlsh.org/CommandSource">
   		<synopsis>Executes a script within the current shell</synopsis>
   </command>
   <command name="test" url="http://www.xmlsh.org/CommandTest" >
   		<synopsis>Evaluates expressions</synopsis>
   </command>
   <command name="throw" url="http://www.xmlsh.org/CommandThrow">
   		<synopsis>create and update timestamp of a file</synopsis>
   </command>
   <command name="tie" url="http://www.xmlsh.org/CommandTie">
   		<synopsis>Ties an xquery to a variable for ease of access</synopsis>
   </command>
   <command name="true" url="http://www.xmlsh.org/CommandTrue">
   		<synopsis>returns true</synopsis>
   </command>
   <command name="unset" url="http://www.xmlsh.org/CommandUnset">
		<synopsis>Unsets a variable</synopsis>
   </command>
   <command name="wait" url="http://www.xmlsh.org/CommandWait">
   		<synopsis>wait for jobs to complete</synopsis>
   </command>
   <command name="xaddattribute" url="http://www.xmlsh.org/CommandXaddattribute">
   		<synopsis>Adds or replaces an attribute to nodes matching an expression</synopsis>
   </command>

   <command name="xaddbase" url="http://www.xmlsh.org/CommandXaddbase">
   		<synopsis>Add xml:base attributes to a document</synopsis>
   </command>
   <command name="xbase" url="http://www.xmlsh.org/CommandXbase">
   		<synopsis>extract the base URI from a document or node</synopsis>
   </command>
   <command name="xcat" url="http://www.xmlsh.org/CommandXcat" >
   		<synopsis>concatenate XML documents</synopsis>
   </command>
   <command name="xcmp" url="http://www.xmlsh.org/CommandXcmp">
   		<synopsis>compare text, binary or xml</synopsis>
   </command>
   <command name="xdelete" url="http://www.xmlsh.org/CommandXDelete">
   		<synopsis>Deletes nodes or attributes matching an expression</synopsis>
   </command>
   <command name="xecho" url="http://www.xmlsh.org/CommandXEcho">
   	<synopsis>Prints arguments as xml</synopsis>
   </command>
   <command name="xed" url="http://www.xmlsh.org/CommandXed">
   		<synopsis>xpath based xml edit</synopsis>
   </command>
   <command name="xfile" url="http://www.xmlsh.org/CommandXfile">
   		<synopsis>file name manipulations</synopsis>
   </command>
   <command name="xgetopts" url="http://www.xmlsh.org/CommandXgetopts">
   		<synopsis>parse arguments into xml</synopsis>
   </command>
      <command name="xgrep" url="http://www.xmlsh.org/CommandXgrep">
	  <synopsis>Search for patterns in files</synopsis>
	</command>
   <command name="xidentity" url="http://www.xmlsh.org/CommandXidentity">
   		<synopsis>Outputs an identical copy of the input</synopsis>
   </command>
   <command name="xinclude" url="http://www.xmlsh.org/CommandXinclude">
   		<synopsis>runs xinclude preprocessing on an xml file</synopsis>
   </command>
   <command name="xls" url="http://www.xmlsh.org/CommandXls" >
   		<synopsis>list files in a directory</synopsis>
   </command>
   <command name="xml2csv" url="http://www.xmlsh.org/CommandXml2csv">
   		<synopsis>convert xml to csv</synopsis>
   </command>
   <command name="xmlsh" url="http://www.xmlsh.org/CommandXmlsh">
   		  		<synopsis>Run a sub shell</synopsis>
   </command>
   <command name="xmove" url="http://www.xmlsh.org/CommandXmove">
   	<synopsis>Rename/Move xml files to a new filename based on an expression in the file</synopsis>
   </command>
   <command name="xpath" url="http://www.xmlsh.org/CommandXpath">
   		<synopsis>run xpath version 2.0 expressions on an xml document</synopsis>
   </command>
   <command name="xproperties" url="http://www.xmlsh.org/CommandXproperties" >
   		<synopsis>parse or generate Java Properties files</synopsis>
   </command>
   <command name="xpwd" url="http://www.xmlsh.org/CommandXpwd">
   		<synopsis>prints the current directory</synopsis>
   </command>
   <command name="xquery" url="http://www.xmlsh.org/CommandXquery" >
   		<synopsis>runs xquery scripts on an xml document</synopsis>
   </command>
   <command name="xread" url="http://www.xmlsh.org/CommandXread">
   		<synopsis>Reads an XML document into a variable</synopsis>
   </command>
   <command name="return" url="http://www.xmlsh.org/CommandReturn">
   		<synopsis>Returns from a function</synopsis>
   </command>
   <command name="xslt" url="http://www.xmlsh.org/CommandXslt" >
   	<synopsis>runs xslt version 2.0 transformations</synopsis>
   </command>
   <command name="xslt1" url="http://www.xmlsh.org/CommandXslt1">
   	<synopsis>runs xslt version 1.0 transformations</synopsis>
   </command>
   <command name="xsplit" url="http://www.xmlsh.org/CommandXsplit">
   <synopsis>splits an XML file into multiple files</synopsis>
   </command>
   <command name="xsql" url="http://www.xmlsh.org/CommandXsql">
   		<synopsis>query a relational database using jdbc</synopsis>
   </command>
   <command name="xtee" url="http://www.xmlsh.org/CommandXtee">
   	<synopsis>copies input to multiple outputs</synopsis>
   </command>
   <command name="xtype" url="http://www.xmlsh.org/CommandXtype">
   		<synopsis>prints the XDM type of each argument</synopsis>
   </command>
   <command name="xuri" url="http://www.xmlsh.org/CommandXuri">
   		<synopsis>create and disect URI's</synopsis>
   </command>
   <command name="xvalidate" url="http://www.xmlsh.org/CommandXvalidate">
   	<synopsis>validate a document against a schema</synopsis>
   </command>
   <command name="xversion" url="http://www.xmlsh.org/CommandXversion">
   	<synopsis>prints the shell version</synopsis>
   </command>
   <command name="xwhich" url="http://www.xmlsh.org/CommandXwhich">
   	<synopsis>Locates commands</synopsis>
   </command>
</commands>
