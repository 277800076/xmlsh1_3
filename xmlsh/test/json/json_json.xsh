import commands json=json
echo json:json( '{ "a" : 1.5 , "b" : [ 1 , true , null , "xyz" ] }' )
