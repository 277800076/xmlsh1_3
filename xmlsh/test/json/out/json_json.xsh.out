{"b":[1,true,null,"xyz"],"a":1.5}
